File: ios/QBSheetLiveKit/Sources/QBSheetLiveKit/Protocol/Bootstrap.swift
import Foundation

/// The QBSheet Live bootstrap URL — the only thing in a printed QR code.
///
/// ```
/// https://live.qbsheet.com/t/<publicationId>?b=<backend origin>&v=1
/// ```
///
/// Parsed here rather than by whichever entry point received it, because there are four of them —
/// a universal link into the full app, an App Clip invocation, a notification tap, and a saved
/// bootstrap restored at launch — and four parsers would be four chances to accept something the
/// others reject.
///
/// The validation mirrors `packages/qblive-protocol/src/bootstrap.ts` exactly; the two have a shared
/// test corpus in `QBLiveBootstrapTests`.
public struct QBLiveBootstrap: Hashable, Sendable, Codable {
    public let version: Int
    public let publicationId: String
    /// The tournament backend's origin. Validated, no trailing slash.
    public let backendOrigin: URL

    public static let currentVersion = 1
    public static let officialHost = "live.qbsheet.com"

    public enum ParseError: Error, LocalizedError, Equatable {
        case notALink
        case noTournament
        case badPublicationId
        case noBackend
        case futureVersion
        case insecureBackend
        case credentialInBackend
        case backendHasPath
        case tooLong

        public var errorDescription: String? {
            switch self {
            case .notALink: "That is not a QBSheet Live link."
            case .noTournament: "That QBSheet Live link does not name a tournament."
            case .badPublicationId: "That QBSheet Live link does not name a valid tournament."
            case .noBackend: "That QBSheet Live link does not name a tournament server."
            case .futureVersion: "That QBSheet Live link needs a newer version of QBSheet Live."
            case .insecureBackend: "A QBSheet Live tournament server must use HTTPS."
            case .credentialInBackend: "That QBSheet Live link is not safe to open."
            case .backendHasPath: "That QBSheet Live link names an invalid tournament server."
            case .tooLong: "That QBSheet Live link is too long."
            }
        }
    }

    /// A generous ceiling that still fits a scannable printed QR code.
    static let maximumLinkLength = 512
    static let maximumOriginLength = 255

    private static let publicationAlphabet = Set("0123456789bcdfghjkmnpqrstvwxyz")

    public static func isPublicationId(_ value: String) -> Bool {
        value.count == 20 && value.allSatisfy { publicationAlphabet.contains($0) }
    }

    public init(version: Int = QBLiveBootstrap.currentVersion, publicationId: String, backendOrigin: URL) {
        self.version = version
        self.publicationId = publicationId
        self.backendOrigin = backendOrigin
    }

    public init(url: URL) throws {
        let raw = url.absoluteString
        guard raw.count <= Self.maximumLinkLength else { throw ParseError.tooLong }
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http"
        else { throw ParseError.notALink }

        let path = components.percentEncodedPath
        let parts = path.split(separator: "/", omittingEmptySubsequences: true)
        guard parts.count == 2, parts[0] == "t" else { throw ParseError.noTournament }
        let publicationId = String(parts[1]).removingPercentEncoding ?? String(parts[1])
        guard Self.isPublicationId(publicationId) else { throw ParseError.badPublicationId }

        let items = components.queryItems ?? []
        let rawVersion = items.first { $0.name == "v" }?.value
        let version = rawVersion.flatMap(Int.init) ?? Self.currentVersion
        guard version >= 1 else { throw ParseError.notALink }
        guard version <= Self.currentVersion else { throw ParseError.futureVersion }

        guard let backend = items.first(where: { $0.name == "b" })?.value, !backend.isEmpty else {
            throw ParseError.noBackend
        }
        self.init(
            version: version,
            publicationId: publicationId,
            backendOrigin: try Self.validateBackendOrigin(backend)
        )
    }

    /// Validate a backend origin hard enough to hand to `URLSession`.
    ///
    /// Each rule answers a specific way a printed QR code could be hostile. Embedded userinfo is
    /// refused because a credential must never travel in a bootstrap URL, and accepting one would
    /// quietly turn a photographed code into a bearer token. A path is refused because the QBLive
    /// routes are appended to this origin and a base path could redirect them. Plain HTTP is
    /// allowed only on the private ranges, which is what Director's local-only LAN mode is.
    public static func validateBackendOrigin(_ raw: String) throws -> URL {
        guard raw.count <= maximumOriginLength else { throw ParseError.tooLong }
        guard let components = URLComponents(string: raw), let host = components.host, !host.isEmpty else {
            throw ParseError.noBackend
        }
        guard components.user == nil, components.password == nil else { throw ParseError.credentialInBackend }
        guard components.query == nil, components.fragment == nil else { throw ParseError.backendHasPath }
        guard components.path.isEmpty || components.path == "/" else { throw ParseError.backendHasPath }

        let scheme = components.scheme?.lowercased()
        if scheme == "https" {
            // Rebuild rather than reuse, so nothing but scheme, host and port survives.
            var clean = URLComponents()
            clean.scheme = "https"
            clean.host = host
            clean.port = components.port
            guard let url = clean.url else { throw ParseError.noBackend }
            return url
        }
        if scheme == "http", isPrivateHost(host) {
            var clean = URLComponents()
            clean.scheme = "http"
            clean.host = host
            clean.port = components.port
            guard let url = clean.url else { throw ParseError.noBackend }
            return url
        }
        throw ParseError.insecureBackend
    }

    /// Loopback and the private ranges. Deliberately narrow: everything else must be HTTPS.
    static func isPrivateHost(_ hostname: String) -> Bool {
        let host = hostname.trimmingCharacters(in: CharacterSet(charactersIn: "[]")).lowercased()
        if host == "localhost" || host == "::1" || host.hasSuffix(".local") { return true }
        let octets = host.split(separator: ".", omittingEmptySubsequences: false).map { Int($0) }
        guard octets.count == 4, octets.allSatisfy({ ($0 ?? -1) >= 0 && ($0 ?? 256) <= 255 }) else { return false }
        let values = octets.compactMap { $0 }
        guard values.count == 4 else { return false }
        switch (values[0], values[1]) {
        case (127, _), (10, _): return true
        case (192, 168): return true
        case (169, 254): return true
        case (172, 16...31): return true
        default: return false
        }
    }

    /// The public URL a Director prints, rebuilt from the parts.
    public func url(officialHost: String = QBLiveBootstrap.officialHost) -> URL {
        var components = URLComponents()
        components.scheme = "https"
        components.host = officialHost
        components.path = "/t/\(publicationId)"
        components.queryItems = [
            URLQueryItem(name: "b", value: backendOrigin.absoluteString),
            URLQueryItem(name: "v", value: String(version)),
        ]
        // The components above are all valid by construction; the fallback keeps this non-optional
        // for callers rather than pushing an impossible failure into every call site.
        return components.url ?? URL(string: "https://\(officialHost)/t/\(publicationId)")!
    }

    /// The QBLive route for a public document.
    public func endpoint(_ path: String) -> URL {
        backendOrigin
            .appendingPathComponent("qblive")
            .appendingPathComponent("v1")
            .appendingPathComponent("tournaments")
            .appendingPathComponent(publicationId)
            .appendingPathComponent(path)
    }

    /// The WebSocket URL, or nil when the backend does not advertise a stream.
    public func streamURL(capabilities: QBLiveCapabilities) -> URL? {
        guard capabilities.stream else { return nil }
        var components = URLComponents(url: endpoint("stream"), resolvingAgainstBaseURL: false)
        components?.scheme = backendOrigin.scheme == "https" ? "wss" : "ws"
        return components?.url
    }
}

File: packages/qblive-protocol/src/bootstrap.ts
/**
 * The QBSheet Live bootstrap URL: the one thing a QR code contains.
 *
 * ```
 * https://live.qbsheet.com/t/<publicationId>?b=<backend>&v=1
 * ```
 *
 * # What is and is not in here
 *
 * Public routing information only: which tournament, and which server holds it. No management
 * token, no QBTCP credential, no APNs publisher secret, no private tournament token. A QR code is
 * photographed by strangers and reposted; treating it as a bearer credential for anything would be
 * a mistake that cannot be walked back once the code is printed.
 *
 * `live.qbsheet.com` is a *bootstrap*, not a proxy. It tells the client where the tournament's own
 * backend is and then gets out of the way — every byte of tournament data is fetched from `b`.
 */

export const QBLIVE_BOOTSTRAP_VERSION = 1;

/** The official invocation domain. Also the only domain the official App Clip is associated with. */
export const QBLIVE_OFFICIAL_ORIGIN = 'https://live.qbsheet.com';

export interface QbliveBootstrap {
  version: number;
  publicationId: string;
  /** The tournament backend origin, already validated. No trailing slash. */
  backendOrigin: string;
}

export class QbliveBootstrapError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'QbliveBootstrapError';
  }
}

/** A generous ceiling that still keeps a printed QR code inside a scannable version. */
const maxBootstrapUrlLength = 512;
const maxBackendOriginLength = 255;

const publicationIdPattern = /^[0-9bcdfghjkmnpqrstvwxyz]{20}$/;

/**
 * Validate a backend origin hard enough to hand to `fetch`.
 *
 * The rules exist for specific attacks. `javascript:`/`data:`/`file:` are refused because the value
 * arrives from a QR code a stranger printed. Embedded userinfo is refused because a credential must
 * never travel in a bootstrap URL, and accepting one would quietly make the QR a bearer token. A
 * path is refused because the QBLive routes are appended to this origin and a base path would let a
 * crafted value redirect them.
 *
 * Plain HTTP is allowed only for loopback and RFC 1918 addresses, which is what Director's
 * local-only LAN mode actually is: a gym with no internet and a laptop on 192.168.1.20.
 */
export function assertPublicBackendOrigin(
  value: string,
  options: { allowInsecureLan?: boolean } = {},
): string {
  if (typeof value !== 'string' || value.length === 0) {
    throw new QbliveBootstrapError('A QBSheet Live backend address is required.');
  }
  if (value.length > maxBackendOriginLength) {
    throw new QbliveBootstrapError('That QBSheet Live backend address is too long.');
  }
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new QbliveBootstrapError('That QBSheet Live backend address is not a valid URL.');
  }
  if (url.username || url.password) {
    throw new QbliveBootstrapError('A QBSheet Live backend address must not contain a username or password.');
  }
  if (url.search || url.hash) {
    throw new QbliveBootstrapError('A QBSheet Live backend address must not contain a query or fragment.');
  }
  if (url.pathname !== '/' && url.pathname !== '') {
    throw new QbliveBootstrapError('A QBSheet Live backend address must be an origin without a path.');
  }
  if (url.protocol === 'https:') return `${url.protocol}//${url.host}`;
  if (url.protocol === 'http:' && options.allowInsecureLan && isLocalHost(url.hostname)) {
    return `${url.protocol}//${url.host}`;
  }
  throw new QbliveBootstrapError('A QBSheet Live backend address must use HTTPS.');
}

/** Loopback and the private ranges. Deliberately narrow: anything else must be HTTPS. */
export function isLocalHost(hostname: string): boolean {
  const host = hostname.replace(/^\[|\]$/g, '').toLowerCase();
  if (host === 'localhost' || host === '::1' || host.endsWith('.local')) return true;
  const v4 = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(host);
  if (!v4) return false;
  const octets = v4.slice(1).map(Number);
  if (octets.some((octet) => octet > 255)) return false;
  const [a, b] = octets;
  if (a === 127) return true;
  if (a === 10) return true;
  if (a === 192 && b === 168) return true;
  if (a === 172 && b >= 16 && b <= 31) return true;
  if (a === 169 && b === 254) return true;
  return false;
}

export function isPublicationId(value: string): boolean {
  return publicationIdPattern.test(value);
}

/**
 * Build the bootstrap URL a Director prints.
 *
 * The backend origin is percent-encoded into `b` rather than shortened, because a QBLive server can
 * be anywhere and a QBSheet-operated shortener would put QBSheet back in the path of every scan.
 */
export function buildBootstrapUrl(
  bootstrap: Omit<QbliveBootstrap, 'version'> & { version?: number },
  officialOrigin = QBLIVE_OFFICIAL_ORIGIN,
): string {
  if (!isPublicationId(bootstrap.publicationId)) {
    throw new QbliveBootstrapError('That is not a valid QBSheet Live publication identifier.');
  }
  const backend = assertPublicBackendOrigin(bootstrap.backendOrigin, { allowInsecureLan: true });
  const url = new URL(`/t/${bootstrap.publicationId}`, officialOrigin);
  url.searchParams.set('b', backend);
  url.searchParams.set('v', String(bootstrap.version ?? QBLIVE_BOOTSTRAP_VERSION));
  const built = url.toString();
  if (built.length > maxBootstrapUrlLength) {
    throw new QbliveBootstrapError('That QBSheet Live link is too long to encode in a QR code.');
  }
  return built;
}

/**
 * Parse a bootstrap URL.
 *
 * Accepts the official host and a self-hosted Live Web deployment alike: the path and query carry
 * the meaning, and the host only decides whether the official App Clip can be invoked.
 */
export function parseBootstrapUrl(value: string): QbliveBootstrap {
  if (typeof value !== 'string' || value.length === 0 || value.length > maxBootstrapUrlLength) {
    throw new QbliveBootstrapError('That is not a QBSheet Live link.');
  }
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new QbliveBootstrapError('That is not a QBSheet Live link.');
  }
  if (url.protocol !== 'https:' && url.protocol !== 'http:') {
    throw new QbliveBootstrapError('That is not a QBSheet Live link.');
  }
  const match = /^\/t\/([^/]+)\/?$/.exec(url.pathname);
  if (!match) throw new QbliveBootstrapError('That QBSheet Live link does not name a tournament.');
  let publicationId: string;
  try {
    publicationId = decodeURIComponent(match[1]);
  } catch {
    throw new QbliveBootstrapError('That QBSheet Live link does not name a valid tournament.');
  }
  if (!isPublicationId(publicationId)) {
    throw new QbliveBootstrapError('That QBSheet Live link does not name a valid tournament.');
  }
  const rawVersion = url.searchParams.get('v');
  const version = rawVersion === null ? QBLIVE_BOOTSTRAP_VERSION : Number(rawVersion);
  if (!Number.isInteger(version) || version < 1) {
    throw new QbliveBootstrapError('That QBSheet Live link uses an unknown format.');
  }
  if (version > QBLIVE_BOOTSTRAP_VERSION) {
    throw new QbliveBootstrapError('That QBSheet Live link needs a newer version of QBSheet Live.');
  }
  const backend = url.searchParams.get('b');
  if (!backend) throw new QbliveBootstrapError('That QBSheet Live link does not name a tournament server.');
  return {
    version,
    publicationId,
    backendOrigin: assertPublicBackendOrigin(backend, { allowInsecureLan: true }),
  };
}

/** The QBLive route for a public document on a backend origin. */
export function qbliveUrl(backendOrigin: string, publicationId: string, path: string): string {
  const base = backendOrigin.replace(/\/$/, '');
  return `${base}/qblive/v1/tournaments/${encodeURIComponent(publicationId)}/${path.replace(/^\//, '')}`;
}
